### ISSUE TITLE
`TestReplicateAddAndRemove` fails in go1.5beta2

### ISSUE THREAD

**Issue by tamird** [cockroachdb/cockroach#1769]
Looks like empty raft appends being rejected (thus not advancing the commit index). I expected #1689 to have fixed this.

https://gist.github.com/tamird/908fe952b1c9f6856e8c

**Comment by tschottdorf** [comment]
but the log is a pass?

**Comment by tamird** [comment]
Ah, you're right. Fixed.

**Comment by tschottdorf** [comment]
tl;dr: group 2 receives a snapshot (match=22), but then appears to receive log index 25 next (rejected with hint 22). nothing else is logged and the commands don't seem to execute on the node.

This in itself is weird assuming the transport works. Too bad we don't have the `V(6)` log - I'm assuming this isn't easily reproduced? Ignoring that it's already weird that we get this message, and assuming that the next thing the leader does (without any logs) is resubmit and catch up the follower, yes, it could be the missing commit index bump again. This isn't fixed yet. Whenever a follower is caught up with the last log entry of the test, if that entry isn't committed yet, there will be no follow-up append sent by the leader (since they don't track the follower's commit index).
A good testing-only fix is to make sure things are happening in Raft. Submitting bogus commands in the `IsTrueWithin` loop will do, but as @bdarnell suggested we should just add the commit index to the peer info and start sending empty appends. Btw, in vanilla Raft heartbeats do bump the commit index, but we special-case those in a way that loses that information.

**Comment by tamird** [comment]
> Too bad we don't have the V(6) log - I'm assuming this isn't easily reproduced?
Heh, it reproduces pretty reliably unless I try to get the V(6) log, in which case it does not.

OK so your analysis sounds good by why does the follower reject the append at all? I was expecting the synchronous dispatch change to totally prevent that from happening since everything would arrive in-order.

**Comment by tschottdorf** [comment]
that's what I also don't know. It may be the first append after the snapshot, possibly there's a bug there (it may not set the state of the peer correctly based on the snapshot) - if it's just that, it shouldn't matter because we can do with one extra round trip after sending up to 64mb over the wire (still nice to fix at some point).

**Comment by bdarnell** [comment]
This isn't an "empty MsgApp" issue; node 2 is behind and needs new log entries, not just the commit index.

I believe the issue has to do with the way raft avoids retrying the same `MsgApp` more than once per heartbeat interval. These tests run with a manual clock so there are no heartbeats. 

The reason the behavior changed between 1.4 and 1.5 is that there is a flurry of activity when a transaction is committed: the EndTransaction itself is applied, and then the asynchronous goroutines start to resolve the intents. These goroutines are scheduled differently in the two versions. In 1.4, when node 2 rejects an append at index 22, the leader has already appended the resolve operations to its log, so the snapshot is sent for index 24 (in my tests). In 1.5, the snapshot happens first so it still has index 22; when it completes we don't retry the entries that come after 22 because the index hasn't changed.

The simplest fix would be to run heartbeats in these tests, although maybe we should be arranging things so that it would work without waiting out a heartbeat interval.

**Comment by bdarnell** [comment]
It's not a manual clock issue; the test uses a manual clock for database timestamps but raft still uses a real clock for its heartbeats. So I'm not sure what's happening; increasing log verbosity makes the problem disappear.

**Comment by bdarnell** [comment]
I've managed to add enough logging to see more of what's going on without making the problem go away, and it looks like #1689 is insufficient. I'm seeing some messages get reordered in transit.

**Comment by bdarnell** [comment]
Ah, #1689 is irrelevant to this test because it uses a LocalRPCTransport instead of `server.RaftTransport`. I just need to convert LocalRPCTransport to use our RPC server.

**Comment by tamird** [comment]
Heh, I'm wondering if we should lint against imports of `net/rpc`. What do you think?

**Comment by bdarnell** [comment]
We still use the `net/rpc` client. The chance of this recurring is low since A) we don't have many rpc servers and B) most rpc servers aren't as reordering-sensitive as a raft transport.

**Comment by tamird** [comment]
:+1:
